// @verixyz/data-store/build/entities/message.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMessage = exports.createMessageEntity = exports.Message = void 0;
const typeorm_1 = require("typeorm");
const identifier_1 = require("./identifier");
const presentation_1 = require("./presentation");
const credential_1 = require("./credential");
const utils_1 = require("@verixyz/utils");
const uuid_1 = require("uuid");
let Message = class Message extends typeorm_1.BaseEntity {
    setId() {
        if (!this.id) {
            this.id = (0, utils_1.computeEntryHash)(this.raw || (0, uuid_1.v4)());
        }
    }
    setSaveDate() {
        this.saveDate = new Date();
        this.updateDate = new Date();
    }
    setUpdateDate() {
        this.updateDate = new Date();
    }
};
__decorate([
    (0, typeorm_1.BeforeInsert)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], Message.prototype, "setId", null);
__decorate([
    (0, typeorm_1.PrimaryColumn)(),
    __metadata("design:type", String)
], Message.prototype, "id", void 0);
__decorate([
    (0, typeorm_1.BeforeInsert)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], Message.prototype, "setSaveDate", null);
__decorate([
    (0, typeorm_1.BeforeUpdate)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], Message.prototype, "setUpdateDate", null);
__decorate([
    (0, typeorm_1.Column)({ select: false }),
    __metadata("design:type", Date)
], Message.prototype, "saveDate", void 0);
__decorate([
    (0, typeorm_1.Column)({ select: false }),
    __metadata("design:type", Date)
], Message.prototype, "updateDate", void 0);
__decorate([
    (0, typeorm_1.Column)({ nullable: true }),
    __metadata("design:type", Date)
], Message.prototype, "createdAt", void 0);
__decorate([
    (0, typeorm_1.Column)({ nullable: true }),
    __metadata("design:type", Date)
], Message.prototype, "expiresAt", void 0);
__decorate([
    (0, typeorm_1.Column)({ nullable: true }),
    __metadata("design:type", String)
], Message.prototype, "threadId", void 0);
__decorate([
    (0, typeorm_1.Column)(),
    __metadata("design:type", String)
], Message.prototype, "type", void 0);
__decorate([
    (0, typeorm_1.Column)({ nullable: true }),
    __metadata("design:type", String)
], Message.prototype, "raw", void 0);
__decorate([
    (0, typeorm_1.Column)('simple-json', { nullable: true }),
    __metadata("design:type", Object)
], Message.prototype, "data", void 0);
__decorate([
    (0, typeorm_1.Column)('simple-array', { nullable: true }),
    __metadata("design:type", Array)
], Message.prototype, "replyTo", void 0);
__decorate([
    (0, typeorm_1.Column)({ nullable: true }),
    __metadata("design:type", String)
], Message.prototype, "replyUrl", void 0);
__decorate([
    (0, typeorm_1.ManyToOne)((type) => identifier_1.Identifier, (identifier) => identifier.sentMessages, {
        nullable: true,
        cascade: ['insert'],
        eager: true,
        onDelete: 'CASCADE',
    }),
    __metadata("design:type", identifier_1.Identifier)
], Message.prototype, "from", void 0);
__decorate([
    (0, typeorm_1.ManyToOne)((type) => identifier_1.Identifier, (identifier) => identifier.receivedMessages, {
        nullable: true,
        cascade: ['insert'],
        eager: true,
        onDelete: 'CASCADE',
    }),
    __metadata("design:type", identifier_1.Identifier)
], Message.prototype, "to", void 0);
__decorate([
    (0, typeorm_1.Column)('simple-json', { nullable: true }),
    __metadata("design:type", Object)
], Message.prototype, "metaData", void 0);
__decorate([
    (0, typeorm_1.ManyToMany)((type) => presentation_1.Presentation, (presentation) => presentation.messages, {
        cascade: true,
    }),
    (0, typeorm_1.JoinTable)(),
    __metadata("design:type", Array)
], Message.prototype, "presentations", void 0);
__decorate([
    (0, typeorm_1.ManyToMany)((type) => credential_1.Credential, (credential) => credential.messages, { cascade: true }),
    (0, typeorm_1.JoinTable)(),
    __metadata("design:type", Array)
], Message.prototype, "credentials", void 0);
Message = __decorate([
    (0, typeorm_1.Entity)('message')
], Message);
exports.Message = Message;
const createMessageEntity = (args) => {
    const message = new Message();
    message.id = args.id;
    message.threadId = args.threadId;
    message.type = args.type;
    message.raw = args.raw;
    message.data = args.data;
    message.metaData = args.metaData;
    if (args.replyTo) {
        message.replyTo = args.replyTo;
    }
    if (args.replyUrl) {
        message.replyUrl = args.replyUrl;
    }
    if (args.createdAt) {
        message.createdAt = new Date(args.createdAt);
    }
    if (args.expiresAt) {
        message.createdAt = new Date(args.expiresAt);
    }
    if (args.from) {
        const from = new identifier_1.Identifier();
        from.did = args.from;
        message.from = from;
    }
    if (args.to) {
        const to = new identifier_1.Identifier();
        to.did = args.to;
        message.to = to;
    }
    if (args.presentations) {
        message.presentations = args.presentations.map(presentation_1.createPresentationEntity);
    }
    if (args.credentials) {
        message.credentials = args.credentials.map(credential_1.createCredentialEntity);
    }
    return message;
};
exports.createMessageEntity = createMessageEntity;
const createMessage = (args) => {
    const message = {
        id: args.id,
        type: args.type,
        raw: args.raw,
        data: args.data,
        metaData: args.metaData,
    };
    if (args.threadId) {
        message.threadId = args.threadId;
    }
    if (args.replyTo) {
        message.replyTo = args.replyTo;
    }
    if (args.replyTo) {
        message.replyUrl = args.replyUrl;
    }
    if (args.createdAt) {
        message.createdAt = args.createdAt.toISOString();
    }
    if (args.expiresAt) {
        message.expiresAt = args.expiresAt.toISOString();
    }
    if (args.from) {
        message.from = args.from.did;
    }
    if (args.to) {
        message.to = args.to.did;
    }
    if (args.presentations) {
        message.presentations = args.presentations.map((vp) => vp.raw);
    }
    if (args.credentials) {
        message.credentials = args.credentials.map((vc) => vc.raw);
    }
    return message;
};
exports.createMessage = createMessage;
