// @verixyz/key-manager/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryPrivateKeyStore = exports.MemoryKeyStore = exports.AbstractSecretBox = exports.AbstractPrivateKeyStore = exports.AbstractKeyStore = exports.AbstractKeyManagementSystem = exports.KeyManager = void 0;
/**
 * Provides a {@link @verixyz/key-manager#KeyManager | plugin} for the {@link @verixyz/core#Agent}
 * that implements {@link @verixyz/core#IKeyManager} interface
 *
 * @packageDocumentation
 */
var key_manager_1 = require("./key-manager");
Object.defineProperty(exports, "KeyManager", { enumerable: true, get: function () { return key_manager_1.KeyManager; } });
var abstract_key_management_system_1 = require("./abstract-key-management-system");
Object.defineProperty(exports, "AbstractKeyManagementSystem", { enumerable: true, get: function () { return abstract_key_management_system_1.AbstractKeyManagementSystem; } });
var abstract_key_store_1 = require("./abstract-key-store");
Object.defineProperty(exports, "AbstractKeyStore", { enumerable: true, get: function () { return abstract_key_store_1.AbstractKeyStore; } });
var abstract_private_key_store_1 = require("./abstract-private-key-store");
Object.defineProperty(exports, "AbstractPrivateKeyStore", { enumerable: true, get: function () { return abstract_private_key_store_1.AbstractPrivateKeyStore; } });
var abstract_secret_box_1 = require("./abstract-secret-box");
Object.defineProperty(exports, "AbstractSecretBox", { enumerable: true, get: function () { return abstract_secret_box_1.AbstractSecretBox; } });
var memory_key_store_1 = require("./memory-key-store");
Object.defineProperty(exports, "MemoryKeyStore", { enumerable: true, get: function () { return memory_key_store_1.MemoryKeyStore; } });
Object.defineProperty(exports, "MemoryPrivateKeyStore", { enumerable: true, get: function () { return memory_key_store_1.MemoryPrivateKeyStore; } });
