// @verixyz/data-store/build/did-discovery-provider.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProfileDiscoveryProvider = void 0;
class ProfileDiscoveryProvider {
    constructor() {
        this.name = 'profile';
    }
    discoverDid(args, context) {
        return __awaiter(this, void 0, void 0, function* () {
            const matches = [];
            const credentials = yield context.agent.dataStoreORMGetVerifiableCredentialsByClaims({
                where: [
                    { column: 'type', value: ['name'] },
                    { column: 'value', value: [`${args.query}%`], op: 'Like' },
                    { column: 'credentialType', value: ['VerifiableCredential,Profile'] },
                ],
            });
            credentials.forEach((vc) => {
                matches.push({
                    did: vc.verifiableCredential.credentialSubject.id,
                    metaData: {
                        verifiableCredential: vc.verifiableCredential,
                    },
                });
            });
            return {
                provider: this.name,
                matches,
            };
        });
    }
}
exports.ProfileDiscoveryProvider = ProfileDiscoveryProvider;
